// iottalk-js/build-node/context.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _channelPool = _interopRequireDefault(require("./channel-pool"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var _default = function _default() {
  _classCallCheck(this, _default);

  this.url = null;
  this.appID = null;
  this.name = null;
  this.mqttHost = null;
  this.mqttPort = null;
  this.mqttUsername = null;
  this.mqttPassword = null;
  this.mqttClient = null;
  this.IChans = new _channelPool["default"](); // input channel

  this.OChans = new _channelPool["default"](); // output channel

  this.rev = null;
  this.onSignal = null;
  this.onData = null;
  this.onRegister = null;
  this.onDeregister = null;
  this.onConnect = null;
  this.onDisconnect = null;
};

exports["default"] = _default;
